harness/determined/searcher/search_method.py

```diff
@@ -0,0 +1,250 @@
+import dataclasses
+import json
+import pathlib
+import uuid
+from abc import abstractmethod
+from enum import Enum
+from typing import Any, Dict, List, Optional, Set
+
+from determined.common.api import bindings
+from determined.common.experimental import Checkpoint
+
+STATE_FILE = "state"
+
+
+@dataclasses.dataclass
+class SearcherState:
+    failures: Set[uuid.UUID]
+    trial_progress: Dict[uuid.UUID, float]
+    trials_closed: Set[uuid.UUID]
+    trials_created: Set[uuid.UUID]
+    last_event_id: Optional[int] = None
+    experiment_completed: bool = False
+
+    def __init__(self) -> None:
+        self.failures = set()
+        self.trial_progress = {}
+        self.trials_closed = set()
+        self.trials_created = set()
+
+    def to_dict(self) -> Dict[str, Any]:
+        return {
+            "failures": [str(f) for f in self.failures],
+            "trialProgress": {str(k): v for k, v in self.trial_progress.items()},
+            "trialsClosed": [str(t) for t in self.trials_closed],
+            "trialsCreated": [str(t) for t in self.trials_created],
+            "lastEventId": self.last_event_id,
+            "experimentId": self.experiment_id,
+            "experimentCompleted": self.experiment_completed,
+        }
+
+    def from_dict(self, d: Dict[str, Any]) -> None:
+        self.failures = {uuid.UUID(f) for f in d.get("failures", [])}
+        self.trial_progress = {uuid.UUID(k): v for k, v in d.get("trialProgress", {}).items()}
+        self.trials_closed = {uuid.UUID(t) for t in d.get("trialsClosed", [])}
+        self.trials_created = {uuid.UUID(t) for t in d.get("trialsCreated", [])}
+        self.last_event_id = d.get("lastEventId")
+        self.experiment_id = d.get("experimentId")
+        self.experiment_completed = d.get("experimentCompleted", False)
+
+
+class ExitedReason(Enum):
+    ERRORED = "ERRORED"
+    USER_CANCELED = "USER_CANCELED"
+    INVALID_HP = "INVALID_HP"
+    INIT_INVALID_HP = "INIT_INVALID_HP"
```

Comment:
is there a good reason this needs to be distinct from INVALID_HP?

My memory tells me that they were distinct due to architectural reasons at the time, but I don't think that's the case anymore.